auth/authentication.go

```diff
@@ -17,119 +16,163 @@ type Database interface {
 	GetApplicationByToken(token string) (*model.Application, error)
 	GetClientByToken(token string) (*model.Client, error)
 	GetPluginConfByToken(token string) (*model.PluginConf, error)
+
 	GetUserByName(name string) (*model.User, error)
 	GetUserByID(id uint) (*model.User, error)
+	CreateUser(user *model.User) error
+}
+
+// AuthenticationProvider provides authentication methods
+type AuthenticationProvider interface {
+	Authenticate(req *http.Request) (user *model.User, err error)
 }
 
 // Auth is the provider for authentication middleware
 type Auth struct {
-	DB Database
+	authenticators map[string]AuthenticationProvider
+	DB             Database
 }
 
-type authenticate func(tokenID string, user *model.User) (authenticated bool, success bool, userId uint, err error)
+// RegisterAuthenticationProvider registers a new authentication provider
+// use empty string as key to register internal credential manager
+func (a *Auth) RegisterAuthenticationProvider(key string, auth AuthenticationProvider) {
+	if a.authenticators == nil {
+		a.authenticators = make(map[string]AuthenticationProvider)
+	}
+	a.authenticators[key] = auth
+}
 
-// RequireAdmin returns a gin middleware which requires a client token or basic authentication header to be supplied
+func (a *Auth) useDesignatedAuthenticator(ctx *gin.Context, designatedAuthenticator string) (user *model.User, authenticator string, err error) {
+	auth, ok := a.authenticators[designatedAuthenticator]
+	if !ok {
+		return nil, "", ProviderNotFoundError{}
+	}
+	user, err = auth.Authenticate(ctx.Request)
+	if user == nil && err == nil {
+		err = NoAuthProviderError{designatedAuthenticator}
+	}
+	return user, designatedAuthenticator, err
+}
+
+func (a *Auth) getUserInfoFromAuth(ctx *gin.Context) (user *model.User, authenticator string, err error) {
+	if designatedAuthenticator := ctx.GetHeader("X-Gotify-Authenticator"); designatedAuthenticator != "" {
+		return a.useDesignatedAuthenticator(ctx, designatedAuthenticator)
+	}
+
+	internalAuthProvider := a.authenticators[""]
+	if user, err := internalAuthProvider.Authenticate(ctx.Request); user != nil {
```

Comment:
can the internal and external auth providers be merged? That, we don't differentiate between plugin auth and the external auth?

Reply:
I wrote it separately to make sure that the internal provider must be the first to check the request.